Source/bmalloc/libpas/src/libpas/pas_stats.h

```diff
@@ -0,0 +1,311 @@
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_SHIFT    PAS_MIN_MEDIUM_ALIGN_SHIFT
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY          (1ull << PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_SHIFT)
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS  (PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_MAX_SIZE / PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY)
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_SHIFT       PAS_MIN_MARGE_ALIGN_SHIFT
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY             (1ull << PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_SHIFT)
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS     (PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_MAX_SIZE / PAS_STATS_MALLOC_INFO_LOW_GRANULARITY)
+// FIXME: handle larger sizes in a sane way -- powers of two?
+#define PAS_STATS_MALLOC_INFO_OVERSIZE_BUCKETS            (1ull)
+
+#define PAS_STATS_MALLOC_INFO_BUCKET_COUNT_PER_STAT (\
+    PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_OVERSIZE_BUCKETS)
+
+typedef struct {
+    pas_stats_stat_base base;
+    uint64_t total_count;
+    uint64_t count_by_heap_type[pas_stats_heap_type_count];
+    uint64_t count_by_size[PAS_STATS_MALLOC_INFO_BUCKET_COUNT_PER_STAT];
+} pas_stats_malloc_info_data;
+
+#if PAS_ENABLE_STATS
+
+char* pas_stats_malloc_info_log(void* stat_v);
+void pas_stats_malloc_info_record(pas_stats_malloc_info_data* data, pas_stats_heap_type heap_type, size_t size, size_t count);
+
+#define PAS_STATS_FOR_EACH_COUNTER(OP) \
+    OP(malloc_info_bytes, pas_stats_malloc_info_data, pas_stats_malloc_info_log) \
+    OP(malloc_info_allocations, pas_stats_malloc_info_data, pas_stats_malloc_info_log)
+
+// FIXME: in principle it should be possible to automatically generate this via PAS_STATS_FOR_EACH_COUNTER, somehow
+#define PAS_RECORD_STAT_malloc_info_bytes(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, size)
+#define PAS_RECORD_STAT_malloc_info_allocations(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, 1)
+
+#else /* !PAS_ENABLE_STATS */
+#define PAS_RECORD_STAT(name, ...) do { PAS_UNUSED_V(__VA_ARGS__); } while (0)
+#endif /* PAS_ENABLE_STATS */
+
+#define PAS_RECORD_STAT_MALLOC(heap_type, size) \
+    PAS_RECORD_STAT(malloc_info_bytes, heap_type, size); \
+    PAS_RECORD_STAT(malloc_info_allocations, heap_type, size);
+
+/******************************************/
+/* Back to general stat-counter machinery */
+/******************************************/
+
+#if PAS_ENABLE_STATS
+
+void pas_stats_do_accounting_before_recording_stat(void);
+// Assumes g_pas_stats_data.log_lock is held
+void pas_stats_log_all_enabled_stats(void);
+
+// It's OK for the .enabled check here to be non-atomic: we know that all
+// .enabled bits will start initialized to 1, and will be set to 0 at most once.
```

Comment:
Since `.enabled` is of type `bool`, let's be consistent and use `true` and `false` everywhere instead of `1` and `0`.  I see that it is initialized to `true`, but other places in the code sets it to `1` or `0`.  Let's fix all those places too.